### PR TITLE
fix(loongsuite): isolate advice and stream failures

### PR DESCRIPTION
**Pull request by sipercai**
# Description

This PR adds Robin-compatible `hook_advice` and `async_hook_advice`
decorators to the existing `opentelemetry-util-genai` package and applies the
shared fail-open boundary to LiteLLM probe-owned lifecycle work.

The application call and stream iteration remain outside advice. LiteLLM keeps
its existing stream accumulator and returns the original response, chunks, and
application exceptions unchanged. Probe failures in preparation, response
mapping, chunk recording, final callbacks, and lifecycle cleanup are isolated
from the application path.

Streaming context is detached before the stream is returned, while a recording
span may be finalized later from another task or context. `_lifecycle_finalized`
is limited to ordinary synchronous LLM lifecycle idempotency and is excluded
from dataclass equality. This PR does not change or harden the existing
multimodal asynchronous dispatch lifecycle.

This follows the
[OpenTelemetry error-handling requirement](https://opentelemetry.io/docs/specs/otel/error-handling/)
that instrumentation errors must not escape into application code.

The shared API is exposed from `opentelemetry.util.genai`, which LiteLLM already
depends on. Robin synchronization can rewrite
`from opentelemetry.util.genai import hook_advice` to the existing commercial
self-monitor decorator import without adding another runtime package.

Fixes # (N/A)

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature which would cause existing functionality not to work as expected)
- [x] This change requires a documentation update

# How Has This Been Tested?

- [x] Fresh-cache `tox -e precommit`
- [x] `tox -c tox-loongsuite.ini -e py311-test-util-genai`
- [x] `tox -c tox-loongsuite.ini -e py311-test-loongsuite-instrumentation-litellm`
- [x] `tox -c tox-loongsuite.ini -e lint-util-genai`
- [x] `tox -c tox-loongsuite.ini -e lint-loongsuite-instrumentation-litellm`
- [x] Independent review and re-review
- [x] Bounded live DashScope non-streaming, streaming, concurrent, and error-path smoke
- [x] Weaver live-check against the LoongSuite GenAI profile

# Does This PR Require a Core Repo Change?

- [ ] Yes. - Link to PR:
- [x] No.

# Checklist

- [x] Followed the style guidelines of this project
- [x] Changelogs have been updated
- [x] Unit tests have been added
- [x] Documentation has been updated

### Spec and Scope

- Linked issue/spec: no public issue; the revised design was approved before implementation.
- Approved scope: shared fail-open advice in the existing GenAI utility package, plus LiteLLM stream lifecycle isolation and contract tests.
- Changed surface: `opentelemetry-util-genai`, LiteLLM wrappers/tests, and Google ADK Python support metadata.
- Non-goals: Robin self-monitor metrics hardening, universal stream proxies, and multimodal asynchronous lifecycle idempotency.

### Local Checks

| Check | Command | Result | Notes |
| --- | --- | --- | --- |
| Static readiness | `check_loongsuite_pr_readiness.py --repo .` | pass | No errors or warnings. |
| Precommit | fresh-cache `tox -e precommit` | pass | All repository hooks passed. |
| util-genai | `tox -c tox-loongsuite.ini -e py311-test-util-genai` | pass | 288 passed, one credential-gated upload test skipped. |
| LiteLLM | `tox -c tox-loongsuite.ini -e py311-test-loongsuite-instrumentation-litellm` | pass | 67 passed. |
| Lifecycle focus | util-genai `test_handler_lifecycle.py` | pass | 10 passed, including detached and non-recording spans. |
| Lint | util-genai pylint and LiteLLM ruff environments | pass | pylint 10/10; ruff passed. |
| Independent review | two review rounds on the narrowed lifecycle design | pass | Sampled-out stream handling fixed; multimodal hardening explicitly deferred. |
| Privacy scan | branch diff and cassette patterns | pass | No credentials, bearer tokens, private paths, or usernames found. |

### Business Isolation

| Scenario | Status | Evidence |
| --- | --- | --- |
| prepare/start fault | pass | One application call; original response preserved; partial span ended. |
| response mapping/stop fault | pass | Original response preserved; context detached and span ended. |
| fail callback fault | pass | Original application exception object preserved. |
| stream chunk/final callback fault | pass | Original chunk identity/order preserved; callback finalized once. |
| early close/cancellation/GeneratorExit | pass | Close/aclose and cancellation tests preserve application control flow. |
| cross-task or cross-Context stream | pass | Context is detached before return and span finalizes later without token reset leakage. |
| clean sibling after fault | pass | Focused lifecycle and concurrency tests show no leaked parent context. |

### Real E2E Matrix

The live provider evidence was collected on this PR's functional LiteLLM
revision. The final scope-narrowing commit only removes multimodal dispatch
state management and adds ordinary lifecycle regression tests.

| Scenario | Status | Command or Demo | Evidence |
| --- | --- | --- | --- |
| non-streaming | pass | Bounded LiteLLM DashScope smoke | Request completed and emitted an LLM span. |
| streaming | pass | Bounded LiteLLM DashScope stream | Stream completed and finalized an LLM span. |
| concurrency | pass | Bounded concurrent DashScope smoke | Calls completed with distinct trace IDs and no parent/content crossover. |
| error path | pass | Live invalid-model request | Provider error remained visible and the error span finalized. |
| agent/tool/ReAct | N/A | LiteLLM client instrumentation | The integration does not own an agent or tool-execution lifecycle. |
| tool-heavy | N/A | LiteLLM client instrumentation | Tool-call delta accumulation is tested, but this integration does not execute tools. |

### Telemetry and Weaver

| Check | Status | Command or Artifact | Notes |
| --- | --- | --- | --- |
| Span tree / span kinds | pass | In-memory exporter and bounded live smoke | Root LLM spans emitted for each call. |
| Content capture modes | pass | Focused tests and live `NO_CONTENT` smoke | `NO_CONTENT` suppresses message attributes; `SPAN_ONLY` is covered by focused tests. |
| Concurrency isolation | pass | Concurrent live smoke | Calls use distinct trace IDs. |
| Weaver live-check | pass | LoongSuite GenAI advice profile | Live spans produced zero violations. |

### CI

- The current remote revision completed all GitHub checks successfully.
- The next push will rerun CI for the narrowed lifecycle diff.
- Shared util changes intentionally select the full LoongSuite matrix.

### Google ADK Python 3.9 CI

The Python 3.9 Google ADK jobs were dependency-resolution failures rather than
advice or stream regressions. The package metadata and generated matrix now
match Google ADK's Python 3.10+ dependency floor. Python 3.10 oldest/latest
tests, lint, workflow generation, precommit, and wheel metadata checks passed.

### CODE REVIEW

**Review by ralf0131 — APPROVED** [review]
## Review: fix(loongsuite): isolate advice and stream failures

### Summary

New `loongsuite-instrumentation-common` package (23 files, +1310/-9) providing failure-isolated advice helpers and protocol-preserving sync/async stream proxies, following the [OpenTelemetry error-handling requirement](https://opentelemetry.io/docs/specs/otel/error-handling/) that instrumentation errors must not escape into application code.

### Key Changes

**1. `advice.py` (+158 lines) — Failure-isolated callbacks**
- `call_advice` wraps instrumentation callbacks so failures are caught and logged, never propagated to application code
- Application return values, original exceptions, and close results remain unchanged
- Clean separation: application call stays outside the advice boundary

**2. `stream.py` (+355 lines) — Protocol-preserving stream proxies**
- Synchronous and asynchronous stream proxies that preserve the original protocol
- Stream chunks, original exceptions, and context-manager suppression semantics all preserved
- Instrumentation callbacks are best-effort only

**3. LiteLLM integration**
- Routes stream accumulation through `call_advice` — consistent with the new pattern
- `test_stream_isolation.py` (+68) verifies the isolation behavior

**4. Package registration & bootstrap**
- New package registered in CI/lint/test matrix (`tox-loongsuite.ini`, workflow files)
- Bootstrap generator updated to exclude `common` from auto-enabled instrumentations — correct, since it's a shared runtime package
- `generate_loongsuite_bootstrap.py` test coverage updated

### CI Status

✅ **All CI passing** — comprehensive matrix:
- All instrumentation packages × Python 3.8-3.14 + pypy-3.9
- crewai tests passing (Issue #230 environment issue resolved)
- typecheck, spellcheck, precommit, shellcheck, docs, docker-tests
- CLA signed

### Test Coverage

- `test_advice.py` (+100) — advice isolation tests
- `test_stream.py` (+369) — stream proxy tests covering sync/async, chunk handling, exception preservation, context manager semantics
- `test_stream_isolation.py` (+68) — LiteLLM-specific isolation tests
- Bootstrap generator tests updated (+10)

### Verdict

Well-designed shared runtime package that properly isolates instrumentation failures from application code. The stream proxy implementation correctly preserves protocol semantics. Comprehensive test coverage. CI fully green. Approving.

---
*Automated review by github-manager-bot*

File: instrumentation-loongsuite/loongsuite-instrumentation-litellm/src/opentelemetry/instrumentation/litellm/_stream_wrapper.py [hunk 1]
```diff
@@ -362,66 +413,106 @@ async def _wrapped_iteration(self):
...
+                "Error closing LiteLLM async stream: %s",
+                exc,
+                exc_info=True,
+            )
 
     async def _aclose_stream(self) -> None:
+        if self._stream_closed:
+            return
+
+        self._stream_closed = True
```
Comment (ralf0131) [inline instrumentation-loongsuite/loongsuite-instrumentation-litellm/src/opentelemetry/instrumentation/litellm/_stream_wrapper.py]: [Duplicate inline comment — please see the latest version of this comment on the same line. Cleanup in progress.]

File: util/opentelemetry-util-genai/src/opentelemetry/util/genai/extended_advice.py [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+                return func(*args, **kwargs)
+            except Exception as exception:  # pylint: disable=broad-exception-caught
+                _log_advice_failure(
+                    instrumentation_name,
+                    advice_method,
+                    exception,
+                )
+                if throw_exception:
+                    raise
+                return None
```
Comment (ralf0131) [inline util/opentelemetry-util-genai/src/opentelemetry/util/genai/extended_advice.py:60]: [Duplicate inline comment — please see the latest version of this comment on the same line. Cleanup in progress.]

File: util/opentelemetry-util-genai/src/opentelemetry/util/genai/extended_advice.py [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+
+
+def _log_advice_failure(
+    instrumentation_name: str,
+    advice_method: str,
+    exception: Exception,
+) -> None:
+    """Report a suppressed instrumentation failure without leaking logging errors."""
+    try:
+        _logger.debug(
```
Comment (ralf0131) [inline util/opentelemetry-util-genai/src/opentelemetry/util/genai/extended_advice.py]: [Duplicate inline comment — please see the latest version of this comment on the same line. Cleanup in progress.]

File: instrumentation-loongsuite/loongsuite-instrumentation-litellm/src/opentelemetry/instrumentation/litellm/_stream_wrapper.py [hunk 4]
```diff
@@ -270,19 +303,29 @@ def __del__(self):
...
         except Exception as exc:
-            logger.debug("Error finalizing unclosed LiteLLM stream: %s", exc)
+            _debug_safely(
+                "Error finalizing unclosed LiteLLM stream: %s",
+                exc,
+                exc_info=True,
+            )
 
     def _close_stream(self) -> None:
+        if self._stream_closed:
```
Comment (ralf0131) [inline instrumentation-loongsuite/loongsuite-instrumentation-litellm/src/opentelemetry/instrumentation/litellm/_stream_wrapper.py]: [Duplicate inline comment — please see the latest version of this comment on the same line. Cleanup in progress.]

**Review by ralf0131 — COMMENTED** [review]
[Automated cleanup: this review was created during testing of inline comment API. The actual line-level inline comments have been posted directly on the relevant code lines.]

File: instrumentation-loongsuite/loongsuite-instrumentation-litellm/src/opentelemetry/instrumentation/litellm/_stream_wrapper.py [hunk 5]
```diff
@@ -35,6 +36,37 @@
...
+
+@hook_advice("litellm", "stream_finalize")
+def _invoke_stream_callback(
+    callback: Any,
+    span: Any,
+    last_chunk: Any,
+    error: Optional[BaseException],
+) -> None:
+    """Run the external finalization callback as fail-open probe advice."""
+    callback(span, last_chunk, error)
```
Comment (ralf0131) [inline instrumentation-loongsuite/loongsuite-instrumentation-litellm/src/opentelemetry/instrumentation/litellm/_stream_wrapper.py:54]: [Duplicate inline comment — please see the latest version of this comment on the same line. Cleanup in progress.]

File: util/opentelemetry-util-genai/src/opentelemetry/util/genai/extended_advice.py (hunk 2 above)
Comment (ralf0131) [inline util/opentelemetry-util-genai/src/opentelemetry/util/genai/extended_advice.py:60]: [Duplicate inline comment — please see the latest version of this comment on the same line. Cleanup in progress.]

File: util/opentelemetry-util-genai/src/opentelemetry/util/genai/extended_advice.py (hunk 3 above)
Comment (ralf0131) [inline util/opentelemetry-util-genai/src/opentelemetry/util/genai/extended_advice.py]: [Duplicate inline comment — please see the latest version of this comment on the same line. Cleanup in progress.]

**Review by ralf0131 — COMMENTED** [review]
[Automated cleanup: this review was created during testing of inline comment API. The actual line-level inline comments have been posted directly on the relevant code lines.]

File: instrumentation-loongsuite/loongsuite-instrumentation-litellm/src/opentelemetry/instrumentation/litellm/_stream_wrapper.py [hunk 6]
```diff
@@ -220,7 +253,7 @@ def __next__(self):
         try:
             chunk = next(self.stream)
 
-            self._accumulator.record_chunk(chunk)
+            _record_stream_chunk(self._accumulator, chunk)
```
Comment (ralf0131) [inline instrumentation-loongsuite/loongsuite-instrumentation-litellm/src/opentelemetry/instrumentation/litellm/_stream_wrapper.py:243]: [Duplicate inline comment — please see the latest version of this comment on the same line. Cleanup in progress.]

File: instrumentation-loongsuite/loongsuite-instrumentation-litellm/src/opentelemetry/instrumentation/litellm/_stream_wrapper.py [hunk 7]
```diff
@@ -233,7 +266,7 @@ def __next__(self):
             raise
         except Exception as e:
             # Error during streaming
-            logger.debug(f"Error during streaming: {e}")
```
Comment (ralf0131) [inline instrumentation-loongsuite/loongsuite-instrumentation-litellm/src/opentelemetry/instrumentation/litellm/_stream_wrapper.py:236]: 📝 **Note — type hint:** `stream: Iterator` could be `Iterator[Any]` for consistency with the rest of the codebase which uses `Any` for generic type parameters. Minor style suggestion, not a functional issue.

File: util/opentelemetry-util-genai/src/opentelemetry/util/genai/extended_advice.py (hunk 2 above)
Comment (ralf0131) [inline util/opentelemetry-util-genai/src/opentelemetry/util/genai/extended_advice.py:60]: [Duplicate inline comment — please see the latest version of this comment on the same line. Cleanup in progress.]

File: util/opentelemetry-util-genai/src/opentelemetry/util/genai/extended_advice.py (hunk 3 above)
Comment (ralf0131) [inline util/opentelemetry-util-genai/src/opentelemetry/util/genai/extended_advice.py]: [Duplicate inline comment — please see the latest version of this comment on the same line. Cleanup in progress.]

File: util/opentelemetry-util-genai/src/opentelemetry/util/genai/extended_advice.py [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+        if inspect.iscoroutinefunction(func):
+            raise TypeError(
+                "hook_advice cannot decorate coroutine functions; "
+                "use async_hook_advice"
+            )
+
+        @wraps(func)
+        def wrapper(*args: Any, **kwargs: Any) -> Any:
+            try:
+                return func(*args, **kwargs)
```
Comment (ralf0131) [inline util/opentelemetry-util-genai/src/opentelemetry/util/genai/extended_advice.py:49]: 📝 **Note — generator/coroutine rejection:** The decorator explicitly rejects generator functions and coroutine functions (lines 63-74), which is good design — generator bodies execute lazily during iteration, outside the decorator's try/except. Stream iteration must instead be isolated by the owning instrumentation wrapper (as done in `_stream_wrapper.py`). This is a thoughtful guard.

File: util/opentelemetry-util-genai/tests/test_extended_advice.py [hunk 9]
```diff
@@ -0,0 +1,157 @@
+# Copyright The OpenTelemetry Authors
```
Comment (ralf0131) [inline util/opentelemetry-util-genai/tests/test_extended_advice.py:1]: 📝 **Note:** Good test coverage — tests for both `hook_advice` (sync) and `async_hook_advice` (async), including failure isolation and the `throw_exception=True` path. Worth adding a test for the generator function rejection (the `TypeError` raised at line 63-69) if not already covered.

**Review by ralf0131 — COMMENTED** [review]
[Automated cleanup: this review was created during testing of inline comment API. The actual line-level inline comments have been posted directly on the relevant code lines.]

File: util/opentelemetry-util-genai/src/opentelemetry/util/genai/extended_advice.py (hunk 3 above)
Comment (ralf0131) [inline util/opentelemetry-util-genai/src/opentelemetry/util/genai/extended_advice.py]: 📝 **Note — log level:** Using `_logger.debug` for instrumentation failure reporting. This aligns with the OpenTelemetry convention that instrumentation errors should be quiet. If these failures need more visibility in production, consider making the level configurable or using `warning`. But `debug` is a reasonable default since these are expected to be non-critical.

File: util/opentelemetry-util-genai/src/opentelemetry/util/genai/extended_advice.py (hunk 8 above)
Comment (ralf0131) [inline util/opentelemetry-util-genai/src/opentelemetry/util/genai/extended_advice.py:49]: 📝 **Note — success path preserved:** The wrapper correctly returns `func(*args, **kwargs)`, preserving the original function return value on success. This is important for callers that depend on the return value (unlike the failure path which returns `None`).

File: util/opentelemetry-util-genai/src/opentelemetry/util/genai/extended_advice.py (hunk 2 above)
Comment (ralf0131) [inline util/opentelemetry-util-genai/src/opentelemetry/util/genai/extended_advice.py:60]: 📝 **Note — fail-open return value:** `return None` when `throw_exception=False` (the default). This is correct for instrumentation-only callbacks, but worth documenting explicitly in the `hook_advice` docstring that callers must not depend on the return value when `throw_exception=False`. A future caller using the return value would silently get `None` on failure.

File: util/opentelemetry-util-genai/src/opentelemetry/util/genai/extended_advice.py [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+                return await func(*args, **kwargs)
+            except Exception as exception:  # pylint: disable=broad-exception-caught
+                _log_advice_failure(
+                    instrumentation_name,
+                    advice_method,
+                    exception,
+                )
+                if throw_exception:
+                    raise
+                return None
```
Comment (ralf0131) [inline util/opentelemetry-util-genai/src/opentelemetry/util/genai/extended_advice.py:96]: 📝 **Note — async fail-open:** Same `return None` pattern as sync `hook_advice`. Good consistency between sync and async versions. The `throw_exception` flag provides an opt-in escape hatch for callers that need errors to propagate.

File: instrumentation-loongsuite/loongsuite-instrumentation-litellm/src/opentelemetry/instrumentation/litellm/_stream_wrapper.py [hunk 11]
```diff
@@ -35,6 +36,37 @@
...
+
+
+@hook_advice("litellm", "stream_finalize")
+def _invoke_stream_callback(
+    callback: Any,
+    span: Any,
+    last_chunk: Any,
+    error: Optional[BaseException],
+) -> None:
+    """Run the external finalization callback as fail-open probe advice."""
```
Comment (ralf0131) [inline instrumentation-loongsuite/loongsuite-instrumentation-litellm/src/opentelemetry/instrumentation/litellm/_stream_wrapper.py:53]: 💡 **Suggestion:** `callback(span, last_chunk, error)` is wrapped by `@hook_advice("litellm", "stream_finalize")`, which catches `Exception` and returns `None` (fail-open). This is correct for instrumentation-only callbacks — the stream callback should not break application flow. Note that `BaseException` (e.g., `KeyboardInterrupt`) propagates through the decorator, which is the right behavior.

File: instrumentation-loongsuite/loongsuite-instrumentation-litellm/src/opentelemetry/instrumentation/litellm/_stream_wrapper.py [hunk 12]
```diff
@@ -220,7 +253,7 @@ def __next__(self):
         try:
```
Comment (ralf0131) [inline instrumentation-loongsuite/loongsuite-instrumentation-litellm/src/opentelemetry/instrumentation/litellm/_stream_wrapper.py:240]: 📝 **Note:** Extracting `_record_stream_chunk` as a standalone function (instead of inline try/except) improves readability and ensures consistent error handling across sync and async paths (line 124). The fail-open pattern here (swallow `Exception`, log at `debug`) matches the `hook_advice` decorator semantics — good consistency.

File: instrumentation-loongsuite/loongsuite-instrumentation-litellm/src/opentelemetry/instrumentation/litellm/_stream_wrapper.py [hunk 13]
```diff
@@ -362,66 +413,106 @@ async def _wrapped_iteration(self):
...
-        self._close_stream()
-        self._finalize()
+        try:
+            self._close_stream()
+        finally:
+            self._finalize()
 
     def _close_stream(self) -> None:
+        if self._stream_closed:
+            return
```
Comment (ralf0131) [inline instrumentation-loongsuite/loongsuite-instrumentation-litellm/src/opentelemetry/instrumentation/litellm/_stream_wrapper.py:450]: 💡 **Suggestion — code duplication:** The async `_aclose_stream` duplicates the sync close logic from `_close_stream` (same `getattr(self.stream, "close", None)` + `close()` + `_debug_safely` pattern). This is intentional (the `_stream_closed` guard would prevent `_close_stream()` from running), but creates a maintenance risk. Consider extracting the sync close fallback into a shared `_close_sync_fallback()` helper called from both paths.

File: instrumentation-loongsuite/loongsuite-instrumentation-litellm/src/opentelemetry/instrumentation/litellm/_stream_wrapper.py [hunk 14]
```diff
@@ -362,66 +413,106 @@ async def _wrapped_iteration(self):
...
         return False
 
     async def aclose(self):
         """Explicitly close and finalize the async stream."""
-        await self._aclose_stream()
-        self._finalize()
+        try:
+            await self._aclose_stream()
+        finally:
+            self._finalize()
```
Comment (ralf0131) [inline instrumentation-loongsuite/loongsuite-instrumentation-litellm/src/opentelemetry/instrumentation/litellm/_stream_wrapper.py:439]: 📝 **Note — shared close guard:** The `_stream_closed` flag is shared between sync `_close_stream` and async `_aclose_stream`, making them mutually exclusive — only the first called will actually close the stream. This is good idempotency design. Just noting for future readers that calling both `close()` and `aclose()` on the same stream will only close once.

File: instrumentation-loongsuite/loongsuite-instrumentation-litellm/src/opentelemetry/instrumentation/litellm/_stream_wrapper.py (hunk 7 above)
Comment (sipercai) [reply instrumentation-loongsuite/loongsuite-instrumentation-litellm/src/opentelemetry/instrumentation/litellm/_stream_wrapper.py:236]: Updated the annotation to `Iterator[Any]` in 956fd307.

File: util/opentelemetry-util-genai/tests/test_extended_advice.py (hunk 9 above)
Comment (sipercai) [reply util/opentelemetry-util-genai/tests/test_extended_advice.py:1]: This is already covered by `test_hook_advice_rejects_generator_function` and `test_async_hook_advice_rejects_async_generator_function`.

File: util/opentelemetry-util-genai/src/opentelemetry/util/genai/extended_advice.py (hunk 2 above)
Comment (sipercai) [reply util/opentelemetry-util-genai/src/opentelemetry/util/genai/extended_advice.py:60]: Documented the suppressed-failure `None` return contract for both decorators in 956fd307.

File: instrumentation-loongsuite/loongsuite-instrumentation-litellm/src/opentelemetry/instrumentation/litellm/_stream_wrapper.py (hunk 13 above)
Comment (sipercai) [reply instrumentation-loongsuite/loongsuite-instrumentation-litellm/src/opentelemetry/instrumentation/litellm/_stream_wrapper.py:450]: Refactored the duplicated fallback into `_close_sync_stream()` in 956fd307.

File: instrumentation-loongsuite/loongsuite-instrumentation-litellm/src/opentelemetry/instrumentation/litellm/_stream_wrapper.py (hunk 14 above)
Comment (sipercai) [reply instrumentation-loongsuite/loongsuite-instrumentation-litellm/src/opentelemetry/instrumentation/litellm/_stream_wrapper.py:439]: The close guard now records closure only after an actual close succeeds. This preserves idempotency while allowing an async-only source to be closed by a later `aclose()` call; this is covered by `test_async_only_stream_can_aclose_after_sync_close`.

**Comment by ralf0131** [comment]
﻿Thanks for the contribution! The code looks good from a review perspective, but **CI is currently failing**:

| Check | Status |
|-------|--------|
| LoongSuite Test 0 result | ❌ failure |

Could you please fix these CI failures so the PR can be merged? You can find the failure details in the [Actions tab](https://github.com/alibaba/loongsuite-python/actions).

Once CI is green, the PR will be ready for merge.

---
*Automated notification by github-manager-bot*

**Review by ralf0131 — COMMENTED** [review]
﻿### Summary
Refactors LiteLLM stream wrapper to route chunk accumulation and finalization callbacks through shared fail-open `@hook_advice` hooks. Solid refactoring with proper error isolation and idempotent stream lifecycle management.

### Overall Assessment
Good changes that improve robustness of the streaming instrumentation. Code review complete — see findings below. However, **CI is currently failing** (`LoongSuite Test 0 result`). This PR cannot be approved until CI passes. Please fix the CI failure (see the CI reminder comment above).

### Findings
- **[Praise]** `util/opentelemetry-util-genai/src/opentelemetry/util/genai/extended_advice.py` — The `@hook_advice` decorator cleanly wraps callbacks in fail-open logic, isolating instrumentation failures from the application path
- **[Praise]** `_stream_wrapper.py` — The `_debug_safely()` wrapper is a good defensive measure against custom logging handlers that may raise during serialization
- **[Praise]** `_stream_wrapper.py` — The `_stream_closed` idempotent guard correctly prevents double-close issues
- **[Praise]** `_stream_wrapper.py` — The try/finally pattern in `_finalize()` ensures the external callback always runs even when `_close_stream()` throws
- **[Info]** `_stream_wrapper.py:228` — The type change from `Iterator` to `Iterator[Any]` is a minor improvement for type safety
- **[Info]** `_stream_wrapper.py:319` — Changing `Exception` to `BaseException` in `_finalize()` signature allows catching `KeyboardInterrupt`/`SystemExit` — ensure this is intentional and the fail-open advice doesn't swallow them silently

### Improvement Suggestions
- Consider adding a brief docstring to `_debug_safely()` explaining why the try/except is needed (custom logging handlers may raise during `%` formatting)

### CI Status
⚠️ CI is failing: `LoongSuite Test 0 result`. This PR will be approved once CI passes.

---
*Automated review by github-manager-bot*

**Review by ralf0131 — APPROVED** [review]
﻿Thanks for the follow-up commit. CI is now fully green.

### Summary of new changes (commit `dcf2284`)

**CI fix: drop Google ADK Python 3.9 jobs** — aligns the `loongsuite-instrumentation-google-adk` package with Google ADK's own Python 3.10+ runtime floor.

Changes reviewed:
- `pyproject.toml` — `requires-python` bumped from `>=3.9` to `>=3.10`; removed `Python :: 3.9` classifier ✅
- `tox-loongsuite.ini` — removed `py39` from the `google-adk` test env matrix (`py3{9,10,11,12,13}` → `py3{10,11,12,13}`) ✅
- `.github/workflows/loongsuite_test_0.yml` — CI matrix updated to drop Python 3.9 for the `google-adk` job ✅
- `CHANGELOG.md` — added a `Changed` entry documenting the Python version floor alignment ✅

### Verdict

Clean, consistent, and correctly scoped — the Python version floor change is applied uniformly across package metadata, tox config, CI workflow, and changelog. All 30 CI checks passing. CLA signed. No merge conflicts.

The core code changes (advice isolation, stream proxies, LiteLLM integration) were already approved in the previous review. This follow-up resolves the CI failures that were blocking merge. **LGTM — approving.**

---
*Automated review by github-manager-bot*

**Review by ralf0131 — APPROVED** [review]
### Summary
Well-architected PR that introduces `hook_advice` and `async_hook_advice` fail-open decorators in `opentelemetry-util-genai` and refactors LiteLLM stream wrappers to use them. The decorators correctly wrap only probe-owned work (chunk recording, finalization callbacks), keeping application calls and stream iteration outside the boundary per OpenTelemetry error-handling spec.

Key design decisions validated:
- Generators/async-generators are rejected (bodies execute after decorator returns)
- `hook_advice` rejects coroutines, `async_hook_advice` rejects sync functions
- `BaseException` is NOT caught — `KeyboardInterrupt`/`CancelledError` propagate correctly
- Logging failures are also isolated (custom handlers may fail)
- Stream lifecycle: `try/finally` ensures `_finalize` always runs even if `_close_stream` throws; `_stream_closed` flag prevents double-close
- f-string logging replaced with `%s` style (logging best practice)

Test coverage is comprehensive (439 lines across `test_extended_advice.py` and `test_stream_isolation.py`), covering success identity, exception swallowing, strict mode, logging isolation, generator rejection, `BaseException` propagation, and async cancellation. No issues found.

---
*Automated review by github-manager-bot*

**Review by ralf0131 — APPROVED** [review]
﻿### Re-review: Changes since last approval (9d8e0f3 → 14c8bc6)

4 new commits since the previous approval. All changes are hardening and simplification of the advice/stream lifecycle:

**Stream wrapper (`_stream_wrapper.py`):**
- `_stream_closed` boolean → `_finalize_lock = Lock()` for thread-safe finalization idempotency
- Added explicit `GeneratorExit` handling — treated as early successful termination, not an error
- `except Exception` → `except BaseException` — now finalizes the span on any termination, then re-raises
- `_finalize()` simplified: lock-guarded idempotency check, callback invoked in `try/finally` so `last_chunk` is always cleared

**Handler lifecycle (`handler.py`, `extended_handler.py`, `types.py`):**
- New `_lifecycle_finalized` flag on `GenAIInvocation` prevents double finalization
- New `detach_llm_context()` separates context detachment from span ending
- New `abandon_llm()` for best-effort cleanup when `start_llm` fails
- `stop_llm`/`fail_llm` use `try/finally` to guarantee context detach + span end
- Checks `span.is_recording()` before `span.end()` — avoids double-end

**Advice (`extended_advice.py`):**
- Removed `_log_advice_failure` helper — logging inlined for simplicity
- Removed `inspect.isgeneratorfunction`/`iscoroutinefunction` checks — now docstring-only guidance

**Tests:** Added `test_completion_isolation.py` (314 lines) and `test_handler_lifecycle.py` (160 lines) covering lifecycle edge cases.

All changes maintain the fail-open design principles validated in the previous review. CLA signed, CI passing. LGTM.

---
*Automated review by github-manager-bot*

**Review by ralf0131 — APPROVED** [review]
﻿### Re-review: Changes since last approval (14c8bc61 → fe362864)

One new commit since the previous approval: `refactor(genai): defer multimodal lifecycle hardening`. The change is a clean simplification of the multimodal lifecycle state management.

### Severity Breakdown
🔴 Critical: 0 | 🟠 High: 0 | 🟡 Medium: 0 | 🔵 Low: 0 | ℹ️ Info: 1 | ✅ Praise: 2

### Key Observations

| Area | Observation |
|------|-------------|
| 🔄 `extended_handler.py` — lifecycle simplification | Removed the fragile `_lifecycle_finalized` toggle (set True before multimodal call, reset on exception). Now `process_multimodal_stop`/`fail` is a simple boolean check — if handled, return; otherwise fall through to parent's sync path. This is safer: if multimodal processing raises, the exception propagates naturally instead of leaving partial state. |
| 📝 `types.py` — `compare=False` | Correctly excludes `_lifecycle_finalized` from dataclass equality comparisons — it's internal runtime state, not a semantic property. |
| ✅ Test coverage | Replaced `test_multimodal_dispatch_failure_allows_finalize_retry` with two focused tests: `test_detached_non_recording_span_is_finalized_once` (verifies idempotent finalization via metric call count) and `test_detached_stream_span_can_be_finalized_by_extended_handler` (verifies detached span context cleanup). Good edge-case coverage. |

### Positive
- Removing the `BaseException` catch is the right call — swallowing all exceptions to manage internal state was an anti-pattern
- The `pylint: disable=unexpected-keyword-arg` comments are appropriate since the `method` param is a runtime convention

ℹ️ **Info:** The CHANGELOG wording change ("synchronous probe finalization") clarifies scope — the hardening now applies to the sync path, with async/streaming handled separately.

---
*Automated review by github-manager-bot*
